api/osf_groups/views.py

```diff
@@ -0,0 +1,224 @@
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        IsGroupManager,
+    )
+    required_read_scopes = [CoreScopes.OSF_GROUPS_READ]
+    required_write_scopes = [CoreScopes.OSF_GROUPS_WRITE]
+    model_class = apps.get_model('osf.OSFGroup')
+
+    parser_classes = (JSONAPIMultipleRelationshipsParser, JSONAPIMultipleRelationshipsParserForRegularJSON,)
+    serializer_class = OSFGroupDetailSerializer
+    view_category = 'osf_groups'
+    view_name = 'group-detail'
+    ordering = ('-modified', )
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def get_object(self):
+        return self.get_osf_group()
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        auth = get_user_auth(self.request)
+        instance.remove_group(auth=auth)
+
+
+class OSFGroupMemberBaseView(JSONAPIBaseView, OSFGroupMixin):
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        IsGroupManager,
+    )
+    required_read_scopes = [CoreScopes.OSF_GROUPS_READ]
+    required_write_scopes = [CoreScopes.OSF_GROUPS_WRITE]
+
+    model_class = apps.get_model('osf.OSFUser')
+    serializer_class = OSFGroupMemberSerializer
+    view_category = 'osf_groups'
+    ordering = ('-modified', )
+
+    def _assert_member_belongs_to_group(self, user):
+        group = self.get_osf_group()
+        # Checking group membership instead of permissions, so unregistered members are
+        # recognized as group members
+        if not group.is_member(user):
+            raise NotFound('{} cannot be found in this OSFGroup'.format(user._id))
+
+    def get_serializer_class(self):
+        if self.request.method == 'PUT' or self.request.method == 'PATCH' or self.request.method == 'DELETE':
+            return OSFGroupMemberDetailSerializer
+        elif self.request.method == 'POST':
+            return OSFGroupMemberCreateSerializer
+        else:
+            return OSFGroupMemberSerializer
+
+    # overrides DestroyAPIView
+    def perform_destroy(self, instance):
+        group = self.get_osf_group()
+        auth = get_user_auth(self.request)
+        try:
+            group.remove_member(instance, auth)
+        except ValueError as e:
+            raise exceptions.ValidationError(detail=e)
+
+
+class OSFGroupMembersList(OSFGroupMemberBaseView, bulk_views.BulkUpdateJSONAPIView, bulk_views.BulkDestroyJSONAPIView, bulk_views.ListBulkCreateJSONAPIView, ListFilterMixin):
+    view_name = 'group-members'
+
+    # Overrides ListAPIView
+    def get_queryset(self):
+        queryset = self.get_queryset_from_request()
+        if is_bulk_request(self.request):
+            user_ids = []
+            for user in self.request.data:
+                try:
+                    user_ids.append(user['id'].split('-')[1])
+                except AttributeError:
```

Comment:
Wouldn't a `KeyError` get raised here? Or is `user['id']` `None` if it isn't provided?

Reply:
The parser has already flattened the request data before it gets here, so `id` is None!

Reply:
Ah I see. It might be a bit more readable to do

```python
try:
    user_id = user['id'].split('-')[1]
except AttributeError:
...
except IndexError:
...
else:
    user_ids.append(user_id)
```

But it's no biggie.

Reply:
done!